### DISCUSSION TITLE
Better support for different middleware functions on different routes

### DISCUSSION THREAD

**Discussion by datteroandrea** [vercel/next.js#45172] [Ideas]
### Describe the feature you'd like to request

I believe that being able of having multiple different middleware functions that work on different routes is needed, the current approach (see [here](https://nextjs.org/docs/messages/nested-middleware)) is not the best for code quality.

### Describe the solution you'd like

We could create a middleware folder where we have multiple different middleware files that handle logic for different routes.
Also multiple different middleware can be run in the same route and there needs to be an order of execution of middlewares.

### Describe alternatives you've considered

The current implementation, however it is not the best.

**Comment by alaato** [comment]
it is definitely more intuitive and makes it easier to make less complex api.

At least for me coming from express.js

**Comment by Oscar-AGUDELO** [comment]
![image](https://github.com/vercel/next.js/assets/89355955/a6fb104e-971d-4995-a13e-7de1fa5ab194)
Hope this can help you ! @datteroandrea

**Reply by alaato** [reply]
how would you handle the response?
for example i would like to show an error page that says that they are un authorized.
returning a response is just gonna give the json back.

**Reply by Oscar-AGUDELO** [reply]
I was planning to deal with error management later but I couldn't help but try it. The example is for a website with several languages but you can adapt it: @alaato 

![image](https://github.com/vercel/next.js/assets/89355955/0ecf4f0e-5df8-45a4-9037-538953a95446)

**Reply by alaato** [reply]
So, making an error route and redirecting there? I thought of it but i preferred to have the error page render on the same route.
Well, we gotta wait for that.
Thank you btw, it's good to see what others think

**Comment by Netail** [comment]
Maybe a solution where you can multiple `middleware.ts` files and will only apply to the routes/subroutes in which the `middleware.ts` is placed.

e.g.
```
app
├── page.tsx
├── layout.tsx
├── middleware.ts
├── stores
│   └── page.tsx
└── dashboard
    ├── page.tsx
    └── middleware.ts
```
where the `middleware.ts` in the root of the app directory applies to all routes and the `middleware.ts` in the dashboard only for the `/dashboard` and subpaths

**Reply by alaato** [reply]
According to the docs, You can't have more than one middleware, only one and it should be inside the root of the project, outside of app.
As far as I know

**Reply by Netail** [reply]
Correct for now, but mine was more of a proposal for the future of middlewares in NextJS 😅

**Reply by alaato** [reply]
Aah, my bad then 😅
